plugins/builder/osbuild.py

```diff
@@ -190,6 +191,59 @@ def from_dict(cls, data: Dict):
         return cls(image_logs, import_logs, init_logs)
 
 
+class OAuth2(requests.auth.AuthBase):
+    class Token:
+        def __init__(self, data):
+            self.data = data["access_token"]
+            self.expires_in = data["expires_in"]
+            self.type = data["token_type"]
+            self.scope = data["scope"]
+
+            self.created = time.time()
+
+        @property
+        def expired(self) -> bool:
+            now = time.time()
+            return now > self.created + self.expires_in
+
+    def __init__(self, cid: str, secret: str, token_url: str) -> None:
+        self.id = cid
+        self.secret = secret
+        self.token_url = token_url
+
+        self.token = None
+
+    @property
+    def token_expired(self) -> bool:
+        return not self.token or self.token.expired
+
+    def fetch_token(self, http: requests.Session):
+        data = {
+            "grant_type": "client_credentials",
+            "client_id": self.id,
+            "client_secret": self.id
+        }
+
+        res = http.post(self.token_url, data=data)
+        if res.status_code != 200:
+            body = res.content.decode("utf-8").strip()
+            msg = f"Failed to authenticate via SSO/OAuth: {body}"
+            raise koji.GenericError(msg) from None
+
+        token_data = res.json()
+        self.token = self.Token(token_data)
+
+    def __call__(self, r: requests.Request):
+        """Called by requests to obtain authorization"""
+
+        # don't add the header if we fetch the token
+        if r.url == self.token_url:
+            return r
+
+        r.headers["authorization"] = "Bearer " + self.token.data
+        return r
+
+
 class Client:
```

Comment:
Does the client get reconstructed each time or are they shared? Unsure how koji plugins work.

Reply:
Don't 100% remember myself, I think it might be per-task, i.e. per request to koji.

Reply:
if that's the case it would be fine, if it's shared might run into some concurrency problems with the token refresh.

Reply:
Good point, let me verify that.